File: frontend/src/app/api/student.ts
import api from './api';

interface StudentScoreProps {
  numCompleted: number;
  numTopScores: number;
  totalScore: number;
  totalAvailableScore: number;
}

interface StudentKuronoGameDataProps {
  worksheetId: number;
  worksheetImage: string;
}

interface NotificationResponseProps {
  notification: string[];
}

const studentApi = api.injectEndpoints({
  endpoints: (build) => ({
    getStudentScore: build.query<StudentScoreProps, null>({
      query: () => ({
        url: 'student/rapid_router_scores/',
        method: 'GET'
      })
    }),
    getStudentKuronoGameData: build.query<StudentKuronoGameDataProps, null>({
      query: () => ({
        url: 'student/kurono_game_data/',
        method: 'GET'
      })
    }),
    updateStudentDetails: build.mutation<
      NotificationResponseProps,
      {
        name: string;
        email: string;
        newPassword: string;
        repeatPassword: string;
        currentPassword: string;
      }
    >({
      query: (body) => ({
        url: 'independent/edit-details/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    }),
    joinSchoolRequest: build.mutation<null, { accessCode: string }>({
      query: (body) => ({
        url: 'student-join-organisation/',
        method: 'POST',
        body: { ...body, classJoinRequest: '' },
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }),
      invalidatesTags: (result) => [{ type: 'studentRequest' }]
    }),
    revokeSchoolRequest: build.mutation<null, { accessCode: string }>({
      query: (body) => ({
        url: 'student-join-organisation/',
        method: 'POST',
        body: { ...body, revokeJoinRequest: '' },
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }),
      invalidatesTags: (result) => [{ type: 'studentRequest' }]
    }),
    isRequestingToJoinSchool: build.query<
      { accessCode: string; isPending: boolean },
      null
    >({
      query: () => ({
        url: 'is-pending-class-request/',
        method: 'GET'
      }),
      providesTags: (result) => [{ type: 'studentRequest' }]
    }),
    updateSchoolStudentDetails: build.mutation<
      NotificationResponseProps,
      {
        newPassword: string;
        repeatPassword: string;
        currentPassword: string;
      }
    >({
      query: (body) => ({
        url: 'school-student/edit-details/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    })
  })
});

export default studentApi;
export const {
  useGetStudentScoreQuery,
  useGetStudentKuronoGameDataQuery,
  useUpdateStudentDetailsMutation,
  useUpdateSchoolStudentDetailsMutation,
  useJoinSchoolRequestMutation,
  useRevokeSchoolRequestMutation,
  useIsRequestingToJoinSchoolQuery
} = studentApi;

File: backend/api/urls/student.py
from django.urls import path

from ..views.student import (
    handle_kurono_game_data,
    handle_rapid_router_scores,
    is_pending_class_request,
    student_join_organisation,
)
from ..views.student.edit_account_details import (
    SchoolStudentEditAccountView,
    change_independent_details,
    delete_independent_account,
)

urlpatterns = [
    path(
        "student/rapid_router_scores/",
        handle_rapid_router_scores,
        name="rr_scores",
    ),
    path(
        "student/kurono_game_data/",
        handle_kurono_game_data,
        name="kurono_game_data",
    ),
    path(
        "independent/edit-details/",
        change_independent_details,
        name="change_independent_details",
    ),
    path(
        "school-student/edit-details/",
        SchoolStudentEditAccountView.as_view(),
        name="school_student_edit_details",
    ),
    path(
        "independent/delete/",
        delete_independent_account,
        name="delete_independent_account",
    ),
    path(
        "student-join-organisation/",
        student_join_organisation,
        name="student_join",
    ),
    path(
        "is-pending-class-request/",
        is_pending_class_request,
        name="is_pending_class_request",
    ),
]
